### PR TITLE
New: [AEA-4039] - Delay 'With Pharmacy' status for 60 minutes

### PR DESCRIPTION
**Pull request by JackSpagnoliNHS**
## Summary

- :sparkles: New Feature

### Details

Adds logic to delay the return of 'With Pharmacy' status on prescriptions that have been released in the last hour, but have not received an NPPT status update, instead returning 'Prescriber Approved' until either 1 hour has passed (Returning 'With Pharmacy but Tracking not Supported'), or an update is received.

Also removes duplicate eslintrc config files, moving to a single root config, sets up prettier-eslint, and sets it as default formatter for json.

### CODE REVIEW

File: packages/enrichPrescriptions/src/statusUpdates.ts [hunk 1]
```diff
@@ -109,20 +117,34 @@ export function applyStatusUpdates(searchsetBundle: Bundle, statusUpdates: Statu
...
     logger.info(`Applying updates for prescription ${prescriptionID}.`)
 
-    const prescriptionUpdate = statusUpdates.prescriptions.filter(p => p.prescriptionID === prescriptionID)[0]
+    const prescriptionUpdate = statusUpdates.prescriptions.filter((p) => p.prescriptionID === prescriptionID)[0]
     if (!prescriptionUpdate || !prescriptionUpdate.onboarded) {
       logger.info(`Supplier of prescription ${prescriptionID} not onboarded. Applying default updates.`)
-      medicationRequests?.forEach(medicationRequest =>
+      medicationRequests?.forEach((medicationRequest) => {
+        if (delayWithPharmacyStatus(medicationRequest)) {
+          logger.info(`Delaying 'With Pharamcy' status for prescription ${prescriptionID}`)
```
Comment (MatthewPopat-NHS) [inline packages/enrichPrescriptions/src/statusUpdates.ts:125]: Let's put the item id on this log too

File: packages/enrichPrescriptions/src/statusUpdates.ts [hunk 2]
```diff
@@ -133,3 +155,44 @@ export function applyStatusUpdates(searchsetBundle: Bundle, statusUpdates: Statu
...
+  return now - updateTime < sixtyMinutes
+}
+
+function getStatusHistoryExtension(medicationRequest: MedicationRequest): Extension | undefined {
+  const STATUS_HISTORY_EXTENSION_URL = "https://fhir.nhs.uk/StructureDefinition/Extension-DM-PrescriptionStatusHistory"
+  return medicationRequest.extension?.find((extension) => extension.url === STATUS_HISTORY_EXTENSION_URL)
+}
+
+export function getStatusDate(statusExtension: Extension): Date | undefined {
+  const dateTime = statusExtension?.extension?.find((extension) => extension?.url === "statusDate")?.valueDateTime
```
Comment (MatthewPopat-NHS) [inline packages/enrichPrescriptions/src/statusUpdates.ts:184]: You're already checking the statusExtension for falsy in the outer function.

File: packages/enrichPrescriptions/src/statusUpdates.ts [hunk 3]
```diff
@@ -133,3 +155,44 @@ export function applyStatusUpdates(searchsetBundle: Bundle, statusUpdates: Statu
...
+  }
+
+  const updateTime = getStatusDate(statusExtension)?.valueOf()
+  const status = getStatus(statusExtension)
+
+  if (!updateTime || !status || status !== "With Pharmacy but Tracking not Supported") {
+    return false
+  }
+
+  const now = new Date().valueOf()
```
Comment (MatthewPopat-NHS) [inline packages/enrichPrescriptions/src/statusUpdates.ts:172]: We should use moment-js to make sure we're working in UTC

File: packages/enrichPrescriptions/src/statusUpdates.ts (hunk 1 above)
Comment (JackSpagnoliNHS) [reply packages/enrichPrescriptions/src/statusUpdates.ts]: f50e933e95502ce9af85fc8306856be8fb4d627b

File: packages/enrichPrescriptions/src/statusUpdates.ts [hunk 4]
```diff
@@ -133,3 +155,44 @@ export function applyStatusUpdates(searchsetBundle: Bundle, statusUpdates: Statu
...
+export function getStatusDate(statusExtension: Extension): Date | undefined {
+  const dateTime = statusExtension?.extension?.find((extension) => extension?.url === "statusDate")?.valueDateTime
+
+  return dateTime ? new Date(dateTime) : undefined
+}
+
+function getStatus(statusExtension: Extension): string | undefined {
+  const VALUE_CODING_SYSTEM = "https://fhir.nhs.uk/CodeSystem/task-businessStatus-nppt"
+
+  return statusExtension?.extension
```
[the remark this answers is not in the document]
Comment (JackSpagnoliNHS) [reply packages/enrichPrescriptions/src/statusUpdates.ts]: 2f71ef74e854dc40df02d9a5fc448ceea71d3a5d

File: packages/enrichPrescriptions/src/statusUpdates.ts [hunk 5]
```diff
@@ -109,20 +120,36 @@ export function applyStatusUpdates(searchsetBundle: Bundle, statusUpdates: Statu
...
+    const prescriptionUpdate = statusUpdates.prescriptions.filter((p) => p.prescriptionID === prescriptionID)[0]
     if (!prescriptionUpdate || !prescriptionUpdate.onboarded) {
       logger.info(`Supplier of prescription ${prescriptionID} not onboarded. Applying default updates.`)
-      medicationRequests?.forEach(medicationRequest =>
+      medicationRequests?.forEach((medicationRequest) => {
+        if (delayWithPharmacyStatus(medicationRequest)) {
+          logger.info(
+            `Delaying 'With Pharmacy' status for prescription ${prescriptionID} line item id ${medicationRequest.id}`
+          )
+          // If the prescription has been in "With Pharmacy" status for less than an hour,
```
Comment (originalphil) [inline packages/enrichPrescriptions/src/statusUpdates.ts:131]: I think this log and comment are a bit misleading. Shouldn't they be '_..."With Pharmacy but Tracking not Supported" status..._'?

Plus, it's True at this point, so no 'if' required.

File: packages/enrichPrescriptions/tests/testStatusUpdate.test.ts [hunk 6]
```diff
@@ -190,4 +196,130 @@ describe("Unit tests for statusUpdate", function () {
...
+
+    const statusUpdates = simpleStatusUpdatesPayload()
+    applyStatusUpdates(requestBundle, statusUpdates)
+
+    // Check that the status has been updated to 'Prescriber Cancelled'
+    expect(medicationRequest.extension![0].extension![0].valueCoding!.code).toEqual("Prescriber Cancelled")
+  })
+
+  describe("Delay WithPharmacy status", () => {
+    it("when the last update is '0002 - With Dispenser', updated less than an hour ago, and missing NPPT updates, set status as 'Prescriber Approved'", async () => {
```
Comment (originalphil) [inline packages/enrichPrescriptions/tests/testStatusUpdate.test.ts:217]: The wording of the test descriptions are a bit confusing here. It isn't immediately obvious that '0002 - With Dispenser' maps to 'With Pharmacy but Tracking not Supported'.

I'd suggest something like "when 'With Pharmacy but Tracking not Supported' is returned by PfP, updated less than an hour ago..."

File: packages/enrichPrescriptions/tests/testStatusUpdate.test.ts [hunk 7]
```diff
@@ -190,4 +196,130 @@ describe("Unit tests for statusUpdate", function () {
 
     expect(requestBundle).toEqual(expectedResponseBundle)
   })
+
+  it("status update does not replace prescriber cancelled status", async () => {
```
Comment (originalphil) [inline packages/enrichPrescriptions/tests/testStatusUpdate.test.ts:200]: Does this one need to be a bit more generic in terms of the status? It implies that Prescriber Cancelled is a special case, when it's just any status other than 'With Pharmacy but Tracking not Supported'.

File: packages/enrichPrescriptions/src/statusUpdates.ts (hunk 5 above)
Comment (JackSpagnoliNHS) [reply packages/enrichPrescriptions/src/statusUpdates.ts]: e3e76eff7306ccb2f4025a4f509860d07431ae7f

File: packages/enrichPrescriptions/tests/testStatusUpdate.test.ts (hunk 6 above)
Comment (JackSpagnoliNHS) [reply packages/enrichPrescriptions/tests/testStatusUpdate.test.ts]: 13da85e71b843341672d94fdcd2b04700c4d409b

**Review by originalphil — APPROVED** [review]

File: packages/enrichPrescriptions/src/statusUpdates.ts [hunk 8]
```diff
@@ -109,20 +120,37 @@ export function applyStatusUpdates(searchsetBundle: Bundle, statusUpdates: Statu
...
-    const prescriptionUpdate = statusUpdates.prescriptions.filter(p => p.prescriptionID === prescriptionID)[0]
+    const prescriptionUpdate = statusUpdates.prescriptions.filter((p) => p.prescriptionID === prescriptionID)[0]
     if (!prescriptionUpdate || !prescriptionUpdate.onboarded) {
       logger.info(`Supplier of prescription ${prescriptionID} not onboarded. Applying default updates.`)
-      medicationRequests?.forEach(medicationRequest =>
+      medicationRequests?.forEach((medicationRequest) => {
+        if (delayWithPharmacyStatus(medicationRequest)) {
+          logger.info(
+            `Delaying 'With Pharmacy but Tracking not Supported' status ` +
+              `for prescription ${prescriptionID} line item id ${medicationRequest.id}`
```
Comment (MatthewPopat-NHS) [inline packages/enrichPrescriptions/src/statusUpdates.ts:130]: The line item id should be the identifier field not the id field.

Comment (JackSpagnoliNHS) [reply packages/enrichPrescriptions/src/statusUpdates.ts]: 909b9e04f41c4cb3bfd0d7c9cee34e4f3d5044ec

**Review by MatthewPopat-NHS — APPROVED** [review]

**Pull request merged**